src/plugins/kibana/server/routes/api/ingest/register_bulk.js

```diff
@@ -0,0 +1,95 @@
+import { Promise } from 'bluebird';
+import { parse, transform } from 'csv';
+import _ from 'lodash';
+import hi from 'highland';
+import { patternToIngest } from '../../../../common/lib/convert_pattern_and_ingest_name';
+import { PassThrough } from 'stream';
+import JSONStream from 'JSONStream';
+
+export function registerBulk(server) {
+  server.route({
+    path: '/api/kibana/{id}/_bulk',
+    method: 'POST',
+    config: {
+      payload: {
+        output: 'stream',
+        maxBytes: 1024 * 1024 * 1024
+      }
+    },
+    handler: function (req, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, req);
+      const indexPattern = req.params.id;
+      const usePipeline = req.query.pipeline === 'true';
+      const delimiter = _.get(req.query, 'delimiter', ',');
+      const responseStream = new PassThrough();
+      const parser = parse({
+        columns: true,
+        auto_parse: true,
+        delimiter: delimiter
+      });
+
+      const csv = req.payload.csv ? req.payload.csv : req.payload;
+      const fileName = req.payload.csv ? csv.hapi.filename : '';
+
+      let currentLine = 2; // Starts at 2 since we parse the header separately
```

Comment:
At the moment the CSV upload UI assumes a header row will be present. It would be cool down the road to give the user a UI driven way to define the column names, but until then I think we can probably punt on support in the API. As a workaround the user could always add a header row to the file.